pkg/cloudprovider/providers/gce/gce_loadbalancer.go

```diff
@@ -88,11 +88,18 @@ func LoadBalancerSrcRanges() []string {
 
 // GetLoadBalancer is an implementation of LoadBalancer.GetLoadBalancer
 func (gce *GCECloud) GetLoadBalancer(clusterName string, service *v1.Service) (*v1.LoadBalancerStatus, bool, error) {
-	loadBalancerName := cloudprovider.GetLoadBalancerName(service)
+	// For backword compatibility we had to fetch the loadBalancerName this way or else a simple assignment for service.Status.LoadBalancer.Name would be enough.
+	var loadBalancerName string
+	if len(service.Status.LoadBalancer.Name) > 0 {
+		loadBalancerName = service.Status.LoadBalancer.Name
```

Comment:
This is more subtle than it looks, I think.

We need to handle the case of "adopting" a service that exists but we have not seen before (e.g. the first time this runs).  That Service will have the old name (legacy) and will not have the status set.  We need to set the status name to the old name.

We need to handle the case of a service that exists but has "lost" status.  Same as above.

We need to handle the case of a service that has the new name (ideal).  In fact, we can't tell the difference between most of these.  If we do not have the status, we have to ask the cloud provider whether the old name exists, and if not ask whether the new name exists.

Worse, GCP LBs are multiple resources, so you have to actually check the earlier-sequence objects.  If the LB is getting an ephemeral IP, you have to check for that.  If it is getting a static IP you might not be able to check by name, so you might have to check the firewall instead.  Why?  Consider the case where you created the IP and/or firewall rule with the old name.  Then you restart the controller into the new version, which restarts the process and tries to use the new name.  If it only checks the ForwardingRule, it will go back and allocate a new IP and firewall with the new name, leaking the old.

So this has to take extra care (and extra comments!!!) to handle the upgrade cleanly.

There is not a clean downgrade.  To do this really cleanly we need a flag that is off by default, and logic to handle either name.  We roll this all out with the flag off, then later, when we're sure we won't roll back beyond this point, flip the flag.

@roberthbailey am I making it harder than it needs to be?